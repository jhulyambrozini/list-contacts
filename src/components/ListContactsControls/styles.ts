import { styled } from 'styled-components'

export const ListControls = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: 6rem 0 4.8rem 0;
`
