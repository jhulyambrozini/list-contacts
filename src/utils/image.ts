const getRandomColor = (): string => {
  const red = Math.floor(Math.random() * 256)
  const green = Math.floor(Math.random() * 256)
  const blue = Math.floor(Math.random() * 256)

  return `rgb(${red},${green},${blue})`
}

export const generateRandomImage = (
  text: string,
  width: number,
  height: number
): string => {
  const canvas = document.createElement('canvas')
  canvas.width = width
  canvas.height = height

  const ctx = canvas.getContext('2d')
  if (!ctx) {
    throw new Error('Canvas 2D context is not supported')
  }

  // Preencher o canvas com uma cor aleatória
  ctx.fillStyle = getRandomColor()
  ctx.fillRect(0, 0, width, height)

  // Configurar o texto no centro
  ctx.fillStyle = 'white'
  ctx.font = '50px Arial'
  ctx.textAlign = 'center'
  ctx.textBaseline = 'middle'

  // Capitalizar a primeira letra da string e colocá-la no centro
  const firstLetter = text.charAt(0).toUpperCase()
  ctx.fillText(firstLetter, width / 2, height / 2)

  // Converter o canvas para um formato de dados URL
  const dataURL = canvas.toDataURL('image/png')
  return dataURL
}

export const validateImageFormat = (file: File): boolean => {
  const allowedFormats = ['image/png', 'image/jpeg']
  return allowedFormats.includes(file.type)
}
