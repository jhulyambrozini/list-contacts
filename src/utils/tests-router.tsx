import { ReactNode } from 'react'
import { BrowserRouter, useNavigate } from 'react-router-dom'
import { render } from '@testing-library/react'

const renderWithRouter = (component: ReactNode) => {
  let navigate: any
  const Wrapper = () => {
    navigate = useNavigate()
    return component
  }

  return {
    ...render(
      <BrowserRouter>
        <Wrapper />
      </BrowserRouter>
    ),
    navigate
  }
}

export default renderWithRouter
