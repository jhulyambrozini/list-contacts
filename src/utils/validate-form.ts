import { FormikProps } from 'formik'
import { ValuesForm } from '../types/Form'

export const getFieldStatus = (
  field: string,
  form: FormikProps<ValuesForm>
) => {
  const isTouched = field in form.touched
  const isInvalid = field in form.errors
  return { isTouched, isInvalid }
}

export const getMessageError = (
  field: string,
  form: FormikProps<ValuesForm>,
  message?: string
) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && isInvalid ? message : false
}

export const getSuccess = (field: string, form: FormikProps<ValuesForm>) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && !isInvalid
}
