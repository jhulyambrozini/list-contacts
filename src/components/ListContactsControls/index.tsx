import { useDispatch, useSelector } from 'react-redux'
import { useNavigate } from 'react-router-dom'

import { Button } from '../../components/Button'

import moreIcon from '../../assets/more-icon.svg'
import sortIcon from '../../assets/sort-icon.svg'

import { ListContactsControlsContainer } from './styles'

import { RootState } from '../../store'
import { sort } from '../../store/reducers/contacts'

const ListContactsControls = () => {
  const { items } = useSelector((state: RootState) => state.contact)
  const dispatch = useDispatch()
  const navigate = useNavigate()

  return (
    <ListContactsControlsContainer>
      <Button.Primary
        type="button"
        padding="1rem 2rem"
        onclick={() => navigate('/new-contact')}
        title="Adicionar contato"
      >
        <Button.Icon alt="Icone de mais" icon={moreIcon} />
        <Button.Label label="ADICIONAR" />
      </Button.Primary>

      {items.length > 0 && (
        <Button.Primary
          type="button"
          padding=".8rem"
          onclick={() => dispatch(sort())}
          title="ordem alfabetica"
        >
          <Button.Icon alt="Icon de A-Z" icon={sortIcon} />
        </Button.Primary>
      )}
    </ListContactsControlsContainer>
  )
}

export default ListContactsControls
