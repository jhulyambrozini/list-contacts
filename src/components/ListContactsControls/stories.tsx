import { Meta } from '@storybook/react'
import mockStore from '../../mocks/redux-store'
import { Provider } from 'react-redux'
import ListContactsControls from '.'

const meta = {
  title: 'components/ListContactsControls',
  component: ListContactsControls,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof ListContactsControls>

export default meta

export const WithListEmpty = () => {
  const store = mockStore({
    contact: {
      items: []
    },
    popUp: {
      isPopupOpen: false
    }
  })
  return (
    <Provider store={store}>
      <ListContactsControls />
    </Provider>
  )
}

export const WithListItems = () => {
  const store = mockStore({
    contact: {
      items: [
        {
          id: '1',
          firstName: 'John',
          lastName: 'Doe',
          email: 'john@example.com',
          tel: '9999-99999',
          image: 'https://placehold.co/90x90/030020/png'
        }
      ]
    },
    popUp: {
      isPopupOpen: false
    }
  })
  return (
    <Provider store={store}>
      <ListContactsControls />
    </Provider>
  )
}
