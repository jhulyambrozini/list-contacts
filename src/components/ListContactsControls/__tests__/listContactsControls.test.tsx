import { MemoryRouter } from 'react-router-dom'
import ListContactsControls from '..'
import { fireEvent, screen } from '@testing-library/react'
import { renderWithProvider } from '../../../utils/tests-redux'
import '@testing-library/jest-dom'
import { vi } from 'vitest'

const renderComponent = () => {
  const { container } = renderWithProvider(
    <MemoryRouter>
      <ListContactsControls />
    </MemoryRouter>,
    {
      preloadedState: {
        contact: {
          items: [
            {
              email: 'exemple@gmail.com',
              id: '1',
              image: '/img.png',
              firstName: 'Kamilly',
              lastName: 'Silva',
              tel: '11 99999-9999'
            },
            {
              email: 'exemple@gmail.co',
              id: '2',
              image: '/img.png',
              firstName: 'Alis',
              lastName: 'Silva',
              tel: '11 99999-8888'
            }
          ]
        }
      }
    }
  )

  return { container }
}

describe('<ListContactsControls />', () => {
  it('deve renderizar botão de orderna conatato somente quando items for maior que 0', () => {
    const { container } = renderComponent()

    const sortButton = screen.getByRole('button', {
      name: /ordem alfabetica/i
    })

    expect(sortButton).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve chamar uma função quando clicar em ordenar', () => {
    const handleClick = vi.fn()
    renderComponent()

    const sortButton = screen.getByRole('button', {
      name: /ordem alfabetica/i
    })

    sortButton.onclick = handleClick

    fireEvent.click(sortButton)

    expect(handleClick).toHaveBeenCalled()
  })

  it('deve chamar uma função quando clicar em adicionar', () => {
    const handleClick = vi.fn()
    renderComponent()

    const addContactButton = screen.getByRole('button', {
      name: /ADICIONAR/i
    })

    addContactButton.onclick = handleClick

    fireEvent.click(addContactButton)

    expect(handleClick).toHaveBeenCalled()
  })
})
