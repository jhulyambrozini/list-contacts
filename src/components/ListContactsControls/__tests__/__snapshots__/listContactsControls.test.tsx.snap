// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ListContactsControls /> > deve renderizar bot√£o de orderna conatato somente quando items for maior que 0 1`] = `
<div
  class="sc-eDnVMP fXkfQf"
>
  <button
    class="sc-bcPKhP sc-grXZZQ fLLMTP eCXYuZ"
    padding="1rem 2rem"
    title="Adicionar contato"
    type="button"
  >
    <img
      alt=""
      src="/src/assets/more-icon.svg"
    />
    <span>
      ADICIONAR
    </span>
  </button>
  <button
    class="sc-bcPKhP sc-grXZZQ dXlDDo eCXYuZ"
    padding=".8rem"
    title="ordem alfabetica"
    type="button"
  >
    <img
      alt=""
      src="/src/assets/sort-icon.svg"
    />
  </button>
</div>
`;
